src/rewrite_clj/node/whitespace.clj

```diff
@@ -68,23 +90,28 @@
          (re-matches #"[\n\r]+" s)]}
   (->NewlineNode s))
 
-(defn- newline?
-  "Check whether a character represents a linebreak."
-  [c]
-  (contains? #{\return \newline} c))
-
 (defn whitespace-nodes
   "Convert a string of whitespace to whitespace/newline nodes."
   [s]
   {:pre [(string? s)
          (re-matches #"(\s|,)+" s)]}
-  (->> (partition-by newline? s)
-       (map
-         (fn [char-seq]
-           (let [s (apply str char-seq)]
-             (if (newline? (first char-seq))
-               (newline-node s)
-               (whitespace-node s)))))))
+  (loop [[c & s' :as s] s
+         acc            []]
+    (cond
+      (empty? s)
+      ,,acc
+
+      (r/linebreak? c)
+      ,,(let [[head tail] (split-with r/linebreak? s)]
+          (recur tail (->> head (apply str) newline-node (conj acc))))
+
+      (r/comma? c)
+      ,,(let [[head tail] (split-with r/comma? s)]
+          (recur tail (->> head (apply str) comma-node (conj acc))))
+
+      :else
+      ,,(let [[head tail] (split-with #(and (r/space? %) (not (r/comma? %))) s)]
+          (recur tail (->> head (apply str) whitespace-node (conj acc)))))))
```

Comment:
Much nicer. My `partition-by` foo is not especially strong and years of banging on C and compilers has brain damaged me into thinking in loops for the most part :stuck_out_tongue: